### PR TITLE
Option to generate sonarqube compatible report

### PR DESCRIPTION
**Pull request by cynthiabaran**
This PR adds the feature of generating a Sonarqube compatible report. To use it, simply add the `--sonarqube` flag when running the CLI.

Further instructions on how to import the report to SonarQube and the formatting of the issues can be found [here](https://docs.sonarqube.org/latest/analysis/generic-issue/).

### CODE REVIEW

**Comment by ajinabraham** [comment]
Thanks for the PR. Can you fix the lint errors?

File: njsscan/njsscan.py [hunk 1]
```diff
@@ -22,9 +22,8 @@ def __init__(self, paths, json, check_controls) -> None:
             'ignore_extensions': conf['ignore_extensions'],
             'ignore_paths': conf['ignore_paths'],
             'ignore_rules': conf['ignore_rules'],
+            'show_progress': show_progress,
         }
-        if not json:
-            self.options['show_progress'] = True
```
Comment (ajinabraham) [inline njsscan/njsscan.py:27]: It was intentional to show progress bar only for njsscan's default CLI output format. This change will enable progress bar only for json and sonarqube output formats and disable it for njsscan CLI. 

Is there a particular reason for enabling progress bar for json and sonarqube output? I guess this output formats will be used in automated CI or programmatic use cases and make no sense to show a progress bar. However, users running njsscan in command-line interface should see the progress bar.

File: njsscan/__main__.py [hunk 2]
```diff
@@ -84,6 +84,65 @@ def json_output(outfile, scan_results):
         return json_output
 
 
+def sonarqube_output(outfile, scan_results):
+    """Sonarqube JSON Output."""
+    sonarqube_issues = []
+    for k, v in scan_results['nodejs'].items():
```
Comment (ajinabraham) [inline njsscan/__main__.py:90]: The scan result also have findings from HTML templates `results['templates']` Is that of no interest?

File: njsscan/__main__.py [hunk 3]
```diff
@@ -84,6 +84,65 @@ def json_output(outfile, scan_results):
...
+    }
+    if secondary_locations:
+        issue['secondaryLocations'] = secondary_locations
+    return issue
+
+
+def translate_to_sonarqube_severity(severity):
+    d = {
+        'ERROR': 'CRITICAL',
+        'WARNING': 'MAJOR',
```
Comment (ajinabraham) [inline njsscan/__main__.py:141]: There is `INFO` as well.

File: njsscan/__main__.py [hunk 4]
```diff
@@ -84,6 +84,65 @@ def json_output(outfile, scan_results):
...
+        'type': 'VULNERABILITY',
+        'severity': translate_to_sonarqube_severity(issue_data['severity']),
+        'primaryLocation': primary_location,
+    }
+    if secondary_locations:
+        issue['secondaryLocations'] = secondary_locations
+    return issue
+
+
+def translate_to_sonarqube_severity(severity):
```
Comment (ajinabraham) [inline njsscan/__main__.py:138]: Is the function `translate_to_sonarqube_severity()` required? Can this mapping 
```
d = {
        'ERROR': 'CRITICAL',
        'WARNING': 'MAJOR',
}
```
be inside `get_sonarqube_issue()` or does it make the function complex?

File: njsscan/__main__.py [hunk 5]
```diff
@@ -126,16 +188,25 @@ def main():
                         action='store_true')
     args = parser.parse_args()
     if args.path:
+        show_progress = False
+        if args.json or args.sonarqube:
+            show_progress = True
+
         scan_results = NJSScan(
```
Comment (ajinabraham) [inline njsscan/__main__.py:195]: This API is used by third-party tools. I do not want to break it for them with this change.

File: njsscan/__main__.py [hunk 6]
```diff
@@ -84,6 +84,65 @@ def json_output(outfile, scan_results):
...
+    }
+    if outfile:
+        with open(outfile, 'w') as of:
+            json.dump(sonarqube_report, of, sort_keys=True,
+                      indent=2, separators=(',', ': '))
+    else:
+        json_output = (json.dumps(sonarqube_report, sort_keys=True,
+                                  indent=2, separators=(',', ': ')))
+        print(json_output)
+        return json_output
```
Comment (ajinabraham) [inline njsscan/__main__.py:105]: This section is redundant. You could simply call `json_output(outfile, sonarqube_report)`

**Comment by cynthiabaran** [comment]
Hey @ajinabraham thanks for the review, I'll publish some fixes soon.

I do have one question, why is the match location described differently for template issues? In nodejs issues, they are described by line and column position. I'm having trouble converting template issues to SonarQube issues.

``` 
{
    "file_path": "tests/assets/templates/true_positives/dust.tl",
    "match_position": [
      12,
      22 
     ]
     ...
}
```

File: njsscan/__main__.py (hunk 5 above)
Comment (cynthiabaran) [reply njsscan/__main__.py:192]: it was my mistake, fixed it!

**Comment by ajinabraham** [comment]
@cynthiabaran njsscan uses 2 engines fot code analysis. A regex based pattern matcher and semantic aware semgrep.

For JavaScript we use semgrep and for HTML templates we use pattern matcher. See an example output format here https://github.com/ajinabraham/libsast#example-usage
Currently Pattern Matcher does not have the ability to return line numbers. Does sonarqube need line numbers?

File: njsscan/__main__.py [hunk 7]
```diff
@@ -87,25 +87,30 @@ def json_output(outfile, scan_results):
...
-            json.dump(sonarqube_report, of, sort_keys=True,
-                      indent=2, separators=(',', ': '))
-    else:
-        json_output = (json.dumps(sonarqube_report, sort_keys=True,
-                                  indent=2, separators=(',', ': ')))
-        print(json_output)
-        return json_output
+    json_output(outfile, sonarqube_report)
+    print(json_output)
+    return json_output
```
Comment (ajinabraham) [inline njsscan/__main__.py:105]: I guess CL 104-105 is not needed?

**Comment by cynthiabaran** [comment]
Yeah, it does require at least a `startLine` number. You can check the required values [here](https://docs.sonarqube.org/latest/analysis/generic-issue/).

**Comment by ajinabraham** [comment]
@cynthiabaran Alright, let's go without template findings for now. I will see if that can be added in future.

**Comment by ajinabraham** [comment]
@cynthiabaran 
I am working on a change to support line numbers form templates (pattern matcher) and also there is a performance bug in semgrep upstream. I will merge this once the other tasks are resolved.

Once again thanks for sending the PR across.

**Comment by cynthiabaran** [comment]
Ok, thanks for addressing the line number issue!

**Comment by ajinabraham** [comment]
Closing in favor of https://github.com/ajinabraham/njsscan/pull/20

**Pull request closed without merging**